Using Mesh doctor generate fractures with internal surfaces already defined

Antoine948398 opened the pull request:
To be merged with the actual branch of @TotoGaz on generate fractures.

AJ0435640 committed:
Add the ability to use external mesh with nternal surfaces already defined

TotoGaz commented:
Thanks, I'll have a look!

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+                if attribute.GetValue(c) in field_values:
+                    nodes = []
+                    for v in range(cell.GetNumberOfPoints()):
+                        node_to_cells[cell.GetPointId(v)] = []
+                        nodes.append(cell.GetPointId(v))
```
```suggestion
                        point_id: int = cell.GetPointId(v)
                        node_to_cells[point_id] = []
                        nodes.append(point_id)
```

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+    if options.policy == "internal_surfaces":
+        node_to_cells = {}
+        face_nodes = []
+        attribute = mesh.GetCellData().GetArray(field)
```
`attribute` is extracted for both `field` and `internal_surfaces`.
Could you use the same (factorized) code (like in lines 96 -> 100)?
Of course `if attribute.GetValue(c) in field_values` line 136 becomes `if attribute[c] in field_values`.

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+        node_to_cells: Mapping[int, Iterable[int]] = build_node_to_cells(mesh, face_nodes)
+    
+    if options.policy == "internal_surfaces":
+        node_to_cells = {}
+        face_nodes = []
```
```suggestion
        node_to_cells: Dict[int, List[int]] = {}
        face_nodes: List[Collection[int]] = []
```

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+        node_to_cells = {}
+        face_nodes = []
+        attribute = mesh.GetCellData().GetArray(field)
+        for c in range( mesh.GetNumberOfCells() ):
```
Can you add a `tqdm` progess bar?

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+                        nodes.append(cell.GetPointId(v))
+                    face_nodes.append(tuple(nodes))
+
+        for c in range( mesh.GetNumberOfCells() ):
```
Can you add a `tqdm` progess bar here as well?

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/parsing/generate_fractures_parsing.py:
```diff
@@ -6,7 +6,8 @@
 __POLICY = "policy"
 __FIELD_POLICY = "field"
-__POLICIES = (__FIELD_POLICY, )
+__INTERNAL_SURFACES_POLICY = "internal_surfaces"
```
Yes, I think that the name `internal_surfaces` is better.

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/parsing/generate_fractures_parsing.py:
```diff
@@ -38,14 +39,14 @@ def fill_subparser(subparsers) -> None:
     p.add_argument('--' + __FIELD_NAME,
                    type=str,
-                   help=f"[string]: If the \"{__FIELD_POLICY}\" {__POLICY} is selected, defines which field will be considered to define the fractures.")
+                   help=f"[string]: If the \"{__FIELD_POLICY}\" {__POLICY} is selected, defines which field will be considered to define the fractures. If the \"{__INTERNAL_SURFACES_POLICY}\" {__POLICY}  is selected, defines the name of the attribute will be considered to identify the fractures. ")
```
Note that there also is a small doc here `src/docs/sphinx/pythonTools/mesh_doctor.rst` that you can use to add other comments if you think it's relevant.

TotoGaz commented on src/coreComponents/python/modules/geosx_mesh_doctor/checks/generate_fractures.py:
```diff
@@ -89,43 +89,64 @@ def build_node_to_cells(mesh: vtkUnstructuredGrid,
+                    face_nodes.append(fn)
+        node_to_cells: Mapping[int, Iterable[int]] = build_node_to_cells(mesh, face_nodes)
+    
+    if options.policy == "internal_surfaces":
```
We should be using `enum`, but I'll manage it after your PR is merged.

AJ0435640 committed:
PR update following comments by @TotoGaz

- Use tqdm when possible
- factorise the "GetArray" by using numpy array for both policies
- Better defintion of dictionnarie/map using strong typed values

Antoine948398 committed:
Merge branch 'TotoGaz/feature/meshDoctorFracture' into feature/amazuyer/mesh_doctor_generatre_fractures_internal_surfaces
